pkg/controller/certificaterequests/venafi/venafi.go

```diff
@@ -0,0 +1,133 @@
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1alpha1"
+	controllerpkg "github.com/jetstack/cert-manager/pkg/controller"
+	"github.com/jetstack/cert-manager/pkg/controller/certificaterequests"
+	crutil "github.com/jetstack/cert-manager/pkg/controller/certificaterequests/util"
+	venafiinternal "github.com/jetstack/cert-manager/pkg/internal/venafi"
+	issuerpkg "github.com/jetstack/cert-manager/pkg/issuer"
+	logf "github.com/jetstack/cert-manager/pkg/logs"
+)
+
+const (
+	CRControllerName = "certificaterequests-issuer-venafi"
+)
+
+type Venafi struct {
+	issuerOptions controllerpkg.IssuerOptions
+	secretsLister corelisters.SecretLister
+	reporter      *crutil.Reporter
+
+	clientBuilder venafiinternal.VenafiClientBuilder
+}
+
+func init() {
+	// create certificate request controller for venafi issuer
+	controllerpkg.Register(CRControllerName, func(ctx *controllerpkg.Context) (controllerpkg.Interface, error) {
+		venafi := NewVenafi(ctx)
+
+		controller := certificaterequests.New(apiutil.IssuerVenafi, venafi)
+
+		c, err := controllerpkg.New(ctx, CRControllerName, controller)
+		if err != nil {
+			return nil, err
+		}
+
+		return c.Run, nil
+	})
+}
+
+func NewVenafi(ctx *controllerpkg.Context) *Venafi {
+	return &Venafi{
+		issuerOptions: ctx.IssuerOptions,
+		secretsLister: ctx.KubeSharedInformerFactory.Core().V1().Secrets().Lister(),
+		reporter:      crutil.NewReporter(ctx.Clock, ctx.Recorder),
+		clientBuilder: venafiinternal.New,
+	}
+}
+
+func (v *Venafi) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerObj cmapi.GenericIssuer) (*issuerpkg.IssueResponse, error) {
+	log := logf.FromContext(ctx, "sign")
+	log = logf.WithRelatedResource(log, issuerObj)
+
+	client, err := v.clientBuilder(cr.Namespace, v.secretsLister, issuerObj)
+	if k8sErrors.IsNotFound(err) {
+		message := "Required secret resource not found"
+
+		v.reporter.Pending(cr, err, "SecretMissing", message)
+		log.Error(err, message)
+
+		return nil, nil
+	}
+
+	if err != nil {
+		message := "Failed to initialise venafi client for signing"
+
+		v.reporter.Pending(cr, err, "VenafiInitError", message)
+		log.Error(err, message)
+
+		return nil, err
+	}
+
+	duration := apiutil.DefaultCertDuration(cr.Spec.Duration)
+
+	certPem, err := client.Sign(cr.Spec.CSRPEM, duration)
+
+	// Check some known error types
+	if err != nil {
+		switch err.(type) {
+
+		case endpoint.ErrCertificatePending:
+			message := "Venafi certificate still in a pending state, the request will be retried"
+
+			v.reporter.Pending(cr, err, "IssuancePending", message)
+			log.Error(err, message)
+			return nil, err
+
+		case endpoint.ErrRetrieveCertificateTimeout:
+			message := "Timed out waiting for venafi certificate, the request will be retried"
+
+			v.reporter.Failed(cr, err, "Timeout", message)
+			log.Error(err, message)
+			return nil, nil
+
+		default:
+			message := "Failed to obtain venafi certificate"
+
+			v.reporter.Pending(cr, err, "RetrieveError", message)
```

Comment:
Should this be `Failed`? Else we're defaulting to retrying on failure...

Reply:
I'd rather fail-safe 😬

Reply:
Makes sense